packages/plugin-hardhat/src/platform/deploy.ts

```diff
@@ -0,0 +1,194 @@
+import type { ethers, ContractFactory } from 'ethers';
+import { CompilerInput, CompilerOutputContract, HardhatRuntimeEnvironment } from 'hardhat/types';
+
+import { BlockExplorerApiKeyClient, SourceCodeLicense } from 'platform-deploy-client';
+import { Network } from 'defender-base-client';
+import { Deployment, DeploymentId, getContractNameAndRunValidation, UpgradesError } from '@openzeppelin/upgrades-core';
+
+import artifactsBuildInfo from '@openzeppelin/upgrades-core/artifacts/build-info.json';
+
+import ERC1967Proxy from '@openzeppelin/upgrades-core/artifacts/@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol/ERC1967Proxy.json';
+import BeaconProxy from '@openzeppelin/upgrades-core/artifacts/@openzeppelin/contracts/proxy/beacon/BeaconProxy.sol/BeaconProxy.json';
+import UpgradeableBeacon from '@openzeppelin/upgrades-core/artifacts/@openzeppelin/contracts/proxy/beacon/UpgradeableBeacon.sol/UpgradeableBeacon.json';
+import TransparentUpgradeableProxy from '@openzeppelin/upgrades-core/artifacts/@openzeppelin/contracts/proxy/transparent/TransparentUpgradeableProxy.sol/TransparentUpgradeableProxy.json';
+import ProxyAdmin from '@openzeppelin/upgrades-core/artifacts/@openzeppelin/contracts/proxy/transparent/ProxyAdmin.sol/ProxyAdmin.json';
+
+import { getNetwork, getPlatformClient } from './utils';
+import { DeployTransaction, PlatformSupportedOptions, UpgradeOptions } from '../utils';
+import debug from '../utils/debug';
+import { getEtherscanAPIConfig } from '../utils/etherscan-api';
+import { getDeployData } from '../utils/deploy-impl';
+import { ContractSourceNotFoundError } from '@openzeppelin/upgrades-core';
+
+const deployableProxyContracts = [
+  ERC1967Proxy,
+  BeaconProxy,
+  UpgradeableBeacon,
+  TransparentUpgradeableProxy,
+  ProxyAdmin,
+];
+
+interface ReducedBuildInfo {
+  _format: string;
+  id: string;
+  solcVersion: string;
+  solcLongVersion: string;
+  input: CompilerInput;
+  output: {
+    contracts: any;
+  };
+}
+
+interface ContractInfo {
+  contractPath: string;
+  contractName: string;
+  buildInfo: ReducedBuildInfo;
+}
+
+type CompilerOutputWithMetadata = CompilerOutputContract & {
+  metadata?: string;
+};
+
+export async function platformDeploy(
+  hre: HardhatRuntimeEnvironment,
+  factory: ContractFactory,
+  opts: UpgradeOptions & PlatformSupportedOptions,
+  ...args: unknown[]
+): Promise<Required<Deployment & DeployTransaction> & DeploymentId> {
+  const client = getPlatformClient(hre);
+
+  const constructorArgs = [...args] as (string | number | boolean)[];
+  const contractInfo = await getContractInfo(hre, factory, { constructorArgs, ...opts });
+  const network = await getNetwork(hre);
+  debug(`Network ${network}`);
+
+  const verifySourceCode = opts.verifySourceCode ?? true;
+  if (verifySourceCode) {
+    await registerEtherscanApiKey(hre, network, client.BlockExplorerApiKey);
```

Comment:
As discussed offline we may want to reconsider doing this automatically.

Reply:
Removed since this is expected to be handled automatically by the Platform